File: frontend/src/types/assessment.ts
export type OrganizationType =
  | "smb"
  | "education"
  | "nonprofit"
  | "government"
  | "consulting"
  | "other";

export type AITechnologyType =
  | "machine_learning"
  | "deep_learning"
  | "nlp"
  | "computer_vision"
  | "rule_based"
  | "generative_ai"
  | "hybrid"
  | "other";

export type VendorType = "in_house" | "vendor" | "both";
export type DevelopmentStatus = "concept" | "development" | "testing" | "production" | "retired";
export type DeploymentStatus = "not_deployed" | "pilot" | "limited" | "full";

export interface Question {
  id: number;
  code: string;
  text: string;
  governance_objective: string | null;
  risk_level: string;
  sort_order: number;
}

export interface Category {
  id: number;
  code: string;
  name: string;
  description: string | null;
  weight: number;
  sort_order: number;
  questions: Question[];
}

export interface OrganizationProfile {
  name: string;
  organization_type: OrganizationType;
  industry?: string;
  country?: string;
  assessment_owner?: string;
  assessment_date?: string;
}

export interface AISystemProfile {
  name: string;
  description?: string;
  primary_purpose?: string;
  technology_type?: AITechnologyType;
  vendor_type?: VendorType;
  development_status?: DevelopmentStatus;
  deployment_status?: DeploymentStatus;
  processes_personal_info?: boolean;
  processes_sensitive_info?: boolean;
  processes_public_data?: boolean;
  processes_employee_data?: boolean;
  processes_customer_data?: boolean;
  processes_student_data?: boolean;
  processes_health_info?: boolean;
  processes_financial_info?: boolean;
  makes_decisions_about_people?: boolean;
  recommends_decisions?: boolean;
  can_materially_affect_individual?: boolean;
  affects_employment?: boolean;
  affects_education?: boolean;
  affects_healthcare?: boolean;
  affects_financial_access?: boolean;
  affects_public_services?: boolean;
  affects_housing?: boolean;
  affects_insurance?: boolean;
  affects_legal_rights?: boolean;
}

export interface ResponseItem {
  question_id: number;
  response_value: number;
  notes?: string;
}

export interface CategoryScore {
  code: string;
  name: string;
  weight_percent: number;
  score: number;
  applicable_questions: number;
}

export interface AssessmentResult {
  overall_score: number;
  readiness_level: string;
  ai_system_risk: string;
  governance_gap: string;
  category_scores: CategoryScore[];
  calculation_trace: Record<string, unknown> | null;
  calculated_at: string;
}

export interface Assessment {
  public_id: string;
  status: string;
  is_demo: boolean;
  organization: OrganizationProfile | null;
  ai_system: AISystemProfile | null;
  responses: ResponseItem[];
  result: AssessmentResult | null;
  created_at: string;
  completed_at: string | null;
}

export const RESPONSE_OPTIONS = [
  { value: 4, label: "Fully implemented" },
  { value: 3, label: "Mostly implemented" },
  { value: 2, label: "Partially implemented" },
  { value: 1, label: "Planned" },
  { value: 0, label: "Not implemented" },
  { value: -1, label: "Not applicable" },
] as const;

export const READINESS_LABELS: Record<string, string> = {
  initial: "Initial",
  developing: "Developing",
  established: "Established",
  advanced: "Advanced",
  leading: "Leading",
};

export const RISK_LABELS: Record<string, string> = {
  low: "Low",
  moderate: "Moderate",
  high: "High",
  critical: "Critical",
};

export const GAP_LABELS: Record<string, string> = {
  low: "Low",
  moderate: "Moderate",
  high: "High",
  critical: "Critical",
};

export const PRIORITY_LABELS: Record<string, string> = {
  critical: "Critical",
  high: "High",
  medium: "Medium",
  low: "Low",
};

export interface Recommendation {
  id?: number;
  category_code: string;
  category_name: string;
  question_id?: number;
  priority: string;
  identified_gap: string;
  recommendation: string;
  why_it_matters: string;
  suggested_action: string;
  responsible_role: string;
  suggested_timeframe: string;
  source_rule_key?: string;
}

export interface RecommendationsResponse {
  total: number;
  recommendations: Recommendation[];
}

export interface RoadmapPhase {
  key: string;
  label: string;
  description: string;
  recommendations: Recommendation[];
}

export interface RoadmapResponse {
  phases: RoadmapPhase[];
}

export interface GovernanceRisk {
  id: number;
  risk_type: string;
  description: string | null;
  likelihood: number;
  impact: number;
  risk_score: number;
  classification: string;
}

export interface RiskMatrixCell {
  likelihood: number;
  impact: number;
  classification: string;
  count: number;
}

export interface AISystemRiskProfile {
  level: string;
  score: number;
  summary: string;
  factors: Array<{ code: string; label: string; present: boolean; weight: number }>;
}

export interface RiskMatrixResponse {
  cells: RiskMatrixCell[];
  risks: GovernanceRisk[];
  ai_system_risk_profile: AISystemRiskProfile;
}

export const RISK_TYPE_LABELS: Record<string, string> = {
  privacy_breach: "Privacy breach",
  algorithmic_bias: "Algorithmic bias",
  incorrect_decisions: "Incorrect decisions",
  lack_of_transparency: "Lack of transparency",
  automation_bias: "Automation bias",
  security_attack: "Security attack",
  data_leakage: "Data leakage",
  model_failure: "Model failure",
  vendor_dependency: "Vendor dependency",
  lack_of_human_oversight: "Lack of human oversight",
  other: "Other",
};

export const CLASSIFICATION_LABELS: Record<string, string> = {
  low: "Low",
  moderate: "Moderate",
  high: "High",
  critical: "Critical",
};

File: backend/app/schemas/assessment.py
"""Assessment API schemas."""

from datetime import date, datetime

from pydantic import BaseModel, Field, field_validator

from app.models.organization import (
    AITechnologyType,
    DeploymentStatus,
    DevelopmentStatus,
    OrganizationType,
    VendorType,
)
from app.models.response import ResponseValue


class OrganizationIn(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    organization_type: OrganizationType
    industry: str | None = Field(default=None, max_length=255)
    country: str | None = Field(default=None, max_length=128)
    assessment_owner: str | None = Field(default=None, max_length=255)
    assessment_date: date | None = None


class AISystemIn(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    description: str | None = None
    primary_purpose: str | None = None
    technology_type: AITechnologyType | None = None
    vendor_type: VendorType | None = None
    development_status: DevelopmentStatus | None = None
    deployment_status: DeploymentStatus | None = None
    processes_personal_info: bool = False
    processes_sensitive_info: bool = False
    processes_public_data: bool = False
    processes_employee_data: bool = False
    processes_customer_data: bool = False
    processes_student_data: bool = False
    processes_health_info: bool = False
    processes_financial_info: bool = False
    makes_decisions_about_people: bool = False
    recommends_decisions: bool = False
    can_materially_affect_individual: bool = False
    affects_employment: bool = False
    affects_education: bool = False
    affects_healthcare: bool = False
    affects_financial_access: bool = False
    affects_public_services: bool = False
    affects_housing: bool = False
    affects_insurance: bool = False
    affects_legal_rights: bool = False


class ProfileIn(BaseModel):
    organization: OrganizationIn
    ai_system: AISystemIn


class ResponseIn(BaseModel):
    question_id: int
    response_value: int
    notes: str | None = None

    @field_validator("response_value")
    @classmethod
    def validate_response_value(cls, value: int) -> int:
        allowed = {item.value for item in ResponseValue}
        if value not in allowed:
            raise ValueError(f"response_value must be one of {sorted(allowed)}")
        return value


class ResponsesIn(BaseModel):
    responses: list[ResponseIn]


class OrganizationOut(OrganizationIn):
    id: int

    model_config = {"from_attributes": True}


class AISystemOut(AISystemIn):
    id: int

    model_config = {"from_attributes": True}


class ResponseOut(BaseModel):
    question_id: int
    response_value: int
    notes: str | None

    model_config = {"from_attributes": True}


class CategoryScoreOut(BaseModel):
    code: str
    name: str
    weight_percent: float
    score: float
    applicable_questions: int


class AssessmentResultOut(BaseModel):
    overall_score: float
    readiness_level: str
    ai_system_risk: str
    governance_gap: str
    category_scores: list[CategoryScoreOut]
    calculation_trace: dict | None
    calculated_at: datetime

    model_config = {"from_attributes": True}


class AssessmentOut(BaseModel):
    public_id: str
    status: str
    is_demo: bool
    organization: OrganizationOut | None = None
    ai_system: AISystemOut | None = None
    responses: list[ResponseOut] = Field(default_factory=list)
    result: AssessmentResultOut | None = None
    created_at: datetime
    completed_at: datetime | None = None

    model_config = {"from_attributes": True}


class AssessmentCreateOut(BaseModel):
    public_id: str
    status: str

    model_config = {"from_attributes": True}
